### PR TITLE
Histogram slider component UI

### PR DESCRIPTION
**Pull request by zlatanpham**
Addresses the UI part of #98. Examples of usage can be found by running `docs`.

### CODE REVIEW

File: src/components/HistogramSlider/Histogram/Histogram.tsx [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    const heightData = data.map(v =>
+      parseFloat((heightPxPerUnit * v).toFixed(2))
+    );
+
+    this.state = {
+      data: heightData
+    };
+  }
+
+  componentWillReceiveProps({ data }: HistogramProps) {
```
Comment (benhinchley) [inline src/components/HistogramSlider/Histogram/Histogram.tsx]: This method has been marked as [`UNSAFE`](https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops)

You can replace it with `static getDerivedStateFromProps()`

File: src/components/HistogramSlider/Histogram/Histogram.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  maskHighlightID: string = `sf-histogram-mask-${maskCount++}`;
+  maskID: string = `sf-histogram-mask-${maskCount++}`;
+  viewBoxWidth: number = this.props.data.length;
+  viewBoxHeight: number =
+    (this.props.heightPx * this.viewBoxWidth) / this.props.widthPx;
+
+  constructor(props: HistogramProps) {
+    super(props);
+
+    const { data } = this.props;
```
Comment (benhinchley) [inline src/components/HistogramSlider/Histogram/Histogram.tsx]: The following can be moved out of the constructor into a `static getDerivedStateFromProps()` method

File: src/components/HistogramSlider/HistogramSlider.tsx [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  HistogramSliderState
+> {
+  static defaultProps = {
+    widthPx: 300,
+    heightPx: 70,
+    colors: {
+      in: "#D7D8D8",
+      out: "#EEEEEE"
+    },
+    InfoRenderComponent: ({ value }: { value: [number, number] }) => (
```
Comment (benhinchley) [inline src/components/HistogramSlider/HistogramSlider.tsx]: I would move this function definition out of here, and create a new function call `DefaultInfoRenderComponent` setting the default prop for `InfoRenderComponent: DefaultInfoRenderComponent`

File: src/components/HistogramSlider/HistogramSlider.tsx [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    this.state = {
+      value: [this.props.value[0], this.props.value[1]]
+    };
+  }
+
+  timeout: number | undefined;
+
+  componentWillReceiveProps(nextProps: HistogramSliderProps) {
```
Comment (benhinchley) [inline src/components/HistogramSlider/HistogramSlider.tsx]: This method has been marked as [`UNSAFE`](https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops)

You can replace it with `static getDerivedStateFromProps()`

File: src/components/HistogramSlider/RangeSlider/RangeSlider.tsx [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+  };
+}
+
+interface RangeSliderState {
+  value: [number, number];
+}
+
+export class RangeSlider extends Component<RangeSliderProps, RangeSliderState> {
+  state: RangeSliderState = {
+    value: [this.props.value[0], this.props.value[1]]
```
Comment (benhinchley) [inline src/components/HistogramSlider/RangeSlider/RangeSlider.tsx]: I would move this into a `static getDerivedStateFromProps` method.

File: src/components/HistogramSlider/RangeSlider/RangeSlider.tsx [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    document.removeEventListener("mouseup", this.clearDocumentEvents);
+    document.removeEventListener("mousemove", this.mouseMoveMin);
+    document.removeEventListener("mousemove", this.mouseMoveMax);
+    document.removeEventListener("touchmove", this.touchMoveMin);
+    document.removeEventListener("touchmove", this.touchMoveMax);
+    document.removeEventListener("touchend", this.clearDocumentEvents);
+    document.removeEventListener("touchcancel", this.clearDocumentEvents);
+  };
+
+  componentWillReceiveProps(nextProps: RangeSliderProps) {
```
Comment (benhinchley) [inline src/components/HistogramSlider/RangeSlider/RangeSlider.tsx]: This method has been marked as [`UNSAFE`](https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops)

You can replace it with `static getDerivedStateFromProps()`

**Comment by zlatanpham** [comment]
Hey @benhinchley I've made the update. Can you take a look? 
Btw, thanks for pointing it out. I should have known the `deprecation` earlier :)

File: src/components/HistogramSlider/RangeSlider/RangeSlider.tsx [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+            tabIndex={0}
+            aria-valuenow={minState}
+            aria-valuemax={max}
+            aria-valuemin={min}
+            aria-disabled="false"
+            onMouseDown={this.triggerMouseMin}
+            onKeyDown={this.handleMinKeydown}
+            onTouchStart={this.triggerTouchMin}
+          />
+          <Button
```
Comment (benhinchley) [inline src/components/HistogramSlider/RangeSlider/RangeSlider.tsx]: I would maybe make this something that a user could override via a prop, like the `InfoRenderComponent`

File: src/components/HistogramSlider/HistogramSlider.tsx [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+const DefaultInfoRenderComponent = ({ value }: { value: [number, number] }) => (
+  <div
+    className={css({
+      marginBottom: "10px",
+      fontSize: "16px",
+      color: "#666666"
+    })}
+  >
+    ${value[0]} AUD - ${value[1]} AUD
```
[the remark this answers is not in the document]
Comment (zlatanpham) [reply src/components/HistogramSlider/HistogramSlider.tsx]: Updated.

File: src/components/HistogramSlider/RangeSlider/RangeSlider.tsx (hunk 7 above)
Comment (zlatanpham) [reply src/components/HistogramSlider/RangeSlider/RangeSlider.tsx]: Updated to allow passing custom drag button via `ButtonRenderComponent` with `focused` state.
```jsx
<HistogramSlider
          ButtonRenderComponent={({ focused }) => (
            <div
              style={{
                width: '20px',
                height: '20px',
                backgroundColor: focused ? 'red' : 'yellow',
              }}
            />
          )}/>
```

**Comment by benhinchley** [comment]
@zlatanpham This is looking good, apologies for not attending to this sooner, I am currently battling with the documentation tool so I can get this running locally.

**Comment by zlatanpham** [comment]
@benhinchley No problem! let me know if you want me take a look at the document problem too.

**Comment by sampotts** [comment]
Hey @benhinchley & @zlatanpham , I'm just looking at [this issue](https://sajari.atlassian.net/browse/SAJARI-622) that's been raised recently and it looks like some of the work in this PR could be used to solve the issue. Any ideas?

**Comment by benhinchley** [comment]
> Hey @benhinchley & @zlatanpham , I'm just looking at [this issue](https://sajari.atlassian.net/browse/SAJARI-622) that's been raised recently and it looks like some of the work in this PR could be used to solve the issue. Any ideas?

Yeah @sampotts This can be used, I for the life of me cannot remember why I didn't merge this in the end.

**Pull request closed without merging**